crates/env_filter/src/parser.rs

```diff
@@ -3,23 +3,42 @@ use log::LevelFilter;
 use crate::Directive;
 use crate::FilterOp;
 
+#[derive(Default, Debug)]
+pub(crate) struct ParseResult {
+    pub(crate) directives: Vec<Directive>,
+    pub(crate) filter: Option<FilterOp>,
+    pub(crate) errors: Vec<String>,
+}
+
+impl ParseResult {
+    fn add_directive(&mut self, directive: Directive) {
+        self.directives.push(directive);
+    }
+
+    fn set_filter(&mut self, filter: FilterOp) {
+        self.filter = Some(filter);
+    }
+
+    fn add_error(&mut self, message: String) {
+        self.errors.push(message);
+    }
+}
+
 /// Parse a logging specification string (e.g: `crate1,crate2::mod3,crate3::x=error/foo`)
 /// and return a vector with log directives.
-pub(crate) fn parse_spec(spec: &str) -> (Vec<Directive>, Option<FilterOp>) {
-    #![allow(clippy::print_stderr)] // compatibility
-
-    let mut dirs = Vec::new();
+pub(crate) fn parse_spec(spec: &str) -> ParseResult {
+    let mut result = ParseResult::default();
 
     let mut parts = spec.split('/');
     let mods = parts.next();
     let filter = parts.next();
     if parts.next().is_some() {
-        eprintln!(
+        result.add_error(format!(
             "warning: invalid logging spec '{}', \
```

Comment:
The `warning:` prefix should be in the caller

Reply:
Ah, you did that in the follow up but I  feel like that should belong in this commit

Reply:
Moved it into the first commit